buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -565,10 +566,59 @@ static ByteBuf encodeString0(ByteBufAllocator alloc, boolean enforceHeap, CharBu
         }
     }
 
+    public static String readUtf8(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof AbstractByteBuf) {
+            return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+        } else {
+            return readUtf8Slow(buf, readerIndex, length);
+        }
+    }
+
+    private static String readUtf8Slow(ByteBuf buf, int readerIndex, int length) {
+        if (buf instanceof WrappedByteBuf) {
+            ByteBuf wrapped;
+            while ((wrapped = buf.unwrap()) != null) {
+                buf = wrapped;
+            }
+            if (buf instanceof AbstractByteBuf) {
+                return readUtf8Fast((AbstractByteBuf) buf, readerIndex, length);
+            }
+        }
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        buf.forEachByte(processor);
+        return processor.toString();
+    }
+
+    // Fast-Path for a AbstractByteBuf
+    static String readUtf8Fast(AbstractByteBuf buf, int readerIndex, int length) {
+        buf.checkIndex(readerIndex, length);
+        UTF8Decoder.UTF8Processor processor = new UTF8Decoder.UTF8Processor(new char[length]);
+        /*
+         * Inline AbstractByteBuf.forEachByte by hand since the JIT doesn't do it it properly
```

Comment:
Heres some more info:
With the hand-inlined micro-optimized version the _entire_ method is recursively inlined into the _method that calls ByteBufUtil.readUtf8_, including the _get call.
This means the _get call can be optimistically inlined by the JIT!
Here's the call chain showing how all the methods are inlined directly into the caller.
![highly inlined call chain](https://i.gyazo.com/02498b1c55c96f33cac61edf89fa1f01.png) 

However, with the 'naive' decoder, almost nothing is inlined
![not inlined call chain](https://i.gyazo.com/0b45d9272d70f2d9beac21e248451b28.png)